Add EVM Gateway node ops section and main page (#1035)
onflow/docs: docs/networks/node-ops/access-onchain-data/evm-gateway/evm-gateway-setup.md

be paid by the EVM Gateway.

The account used for funding gateway Cadence transactions must be a COA, not an EOA. `--coa-address` is configured with the Cadence address
of the COA account and the `--coa-key` must belong to the same account. The `--coinbase` account accrues EVM Gateway fees from EVM client
transactions and can be either an EVM EOA or COA address.

It is acceptable to create a single Cadence account for the COA and use the EVM address associated with that for the COINBASE address.

### Create Flow account to use for COA

If you don't already have a Flow account you will need to create one. 

<Tabs>
<TabItem value="mainnet" label="Mainnet">

   1. Install [Flow Wallet](https://wallet.flow.com/)
   2. Once installed you will be able to copy the wallet address, similar to _0x1844efeb3fef2242_
   3. Obtain account private key from
   <pre>Settings -> Account List -> Choose Main account -> Private Key -> [Password prompt]</pre>
   4. Ensure the wallet is funded from a CEX or other wallet

</TabItem>
<TabItem value="testnet" label="Testnet">
Install [Flow CLI](https://developers.flow.com/tools/flow-cli/install) if not already installed.

> ```bash
> flow keys generate
> This will output something similar to:
> 

Author: ```suggestion
```bash
flow keys generate
This will output something similar to:
```
```

the gateway submits Cadence transactions wrapping EVM transaction payloads to the Flow Access Node the transaction fee for that must
be paid by the EVM Gateway.

The account used for funding gateway Cadence transactions must be a COA, not an EOA. `--coa-address` is configured with the Cadence address
of the COA account and the `--coa-key` must belong to the same account. The `--coinbase` account accrues EVM Gateway fees from EVM client
transactions and can be either an EVM EOA or COA address.

It is acceptable to create a single Cadence account for the COA and use the EVM address associated with that for the COINBASE address.

### Create Flow account to use for COA

If you don't already have a Flow account you will need to create one. 

<Tabs>
<TabItem value="mainnet" label="Mainnet">

   1. Install [Flow Wallet](https://wallet.flow.com/)
   2. Once installed you will be able to copy the wallet address, similar to _0x1844efeb3fef2242_
   3. Obtain account private key from
   <pre>Settings -> Account List -> Choose Main account -> Private Key -> [Password prompt]</pre>
   4. Ensure the wallet is funded from a CEX or other wallet

</TabItem>

> <TabItem value="testnet" label="Testnet">
> 
> ```bash
> flow keys generate
> ```
> 

Author: ```suggestion
<TabItem value="testnet" label="Testnet">
Install [Flow CLI](https://developers.flow.com/tools/flow-cli/install) if not already installed.

```bash
flow keys generate
```

```

---
title: Setting up an EVM Gateway node
sidebar_label: EVM Gateway Setup
sidebar_position: 3
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

This guide is for running the [EVM Gateway](https://github.com/onflow/flow-evm-gateway) node on Flow. The EVM Gateway implements the
[Ethereum JSON-RPC specification](https://ethereum.org/en/developers/docs/apis/json-rpc/) and is the only node type which accepts EVM 
client connections. 

> The EVM Gateway is an unstaked, non-participating edge node which consumes Flow protocol state from the configured Flow Access Node 

Reviewer: why do you say its `non-participating`? IMO, it participates by sending transactions. It doesn't participate in block production directly but does so indirectly.

Author: I'll correct it to say it isn't a participant in block production, would that work?

and persists the indexed EVM state locally to service EVM client requests. It submits EVM transactions it receives into the Flow 

> network, wrapped using a Cadence transaction, and mutating EVM state when executed. Non-mutating RPC methods only query the local state 

Reviewer: ```suggestion
network, wrapped in a Cadence transaction, and mutating EVM state when executed. Non-mutating RPC methods only query the local state 
```

> The EVM Gateway is an unstaked, non-participating edge node which consumes Flow protocol state from the configured Flow Access Node 
> and persists the indexed EVM state locally to service EVM client requests. It submits EVM transactions it receives into the Flow 
> network, wrapped in a Cadence transaction, and mutating EVM state when executed. Non-mutating RPC methods only query the local state 
> index of the gateway and are never forwarded to Access Nodes.

Author: ```suggestion
The EVM Gateway consumes Flow protocol state from the configured Flow Access Node and persists the indexed EVM state locally to service EVM client requests. It submits EVM transactions it receives into the Flow network, wrapped in a Cadence transaction, and mutating EVM state when executed. Non-mutating RPC methods only query the local state index of the gateway and are never forwarded to Access Nodes. It does not participate in the block production process and requires no stake.
```

can be an account created using Metamask or other web3.js wallet, or otherwise can be the EVM address corresponding to the Flow Wallet COA account created above. 

If you haven't already got an EVM address and you have the COA account created by Flow Wallet above then follow the steps below:

* Click top left burger icon to show current profile
* Click 'Enable the path to EVM on Flow' button
* Your EVM account will now be available to use in the left nav account view
* When you switch to that account you can obtain its EVM address

### COA Address and Key

COA address and private key is configured for `--coa-address` & `--coa-key` configuration flags. If running multiple EVM Gateway hosts it is standard to
share the same COA address and key across _n_ hosts.

### Run the gateway

Ensure that the following ENV variables have been set. Add/update as required if your configuration differs from those listed.

```bash
# Set required environment variables
export ACCESS_NODE_GRPC_HOST="access.mainnet.nodes.onflow.org:9000" # or access.devnet.nodes.onflow.org:9000 for testnet
export FLOW_NETWORK_ID="flow-mainnet" # or flow-testnet

> export INIT_CADENCE_HEIGHT="85981134" # 211176670 for testnet

Author: ```suggestion
export INIT_CADENCE_HEIGHT="85981135" # 211176670 for testnet
```

export COINBASE="${EVM_ADDRESS_WITHOUT_0x}"
export COA_ADDRESS="${CADENCE_ACCOUNT_ADDRESS_WITHOUT_0x}"
export COA_KEY="${CADENCE_ACCOUNT_PRIVATE_KEY_WITHOUT_0x}"
export GAS_PRICE="100"
```

<Tabs>
<TabItem value="source-build" label="Run from binary">

**Create EVM Gateway service**

```bash
sudo tee <<EOF >/dev/null /etc/systemd/system/gateway.service
[Unit]
Description=Gateway daemon
After=network-online.target

[Service]
User=$USER
ExecStart=/usr/bin/evm-gateway \
--access-node-grpc-host=$ACCESS_NODE_GRPC_HOST \
--flow-network-id=$FLOW_NETWORK_ID \
--init-cadence-height=$INIT_CADENCE_HEIGHT \
--ws-enabled=true \
--coinbase=$COINBASE \
--coa-address=$COA_ADDRESS \
--coa-key=$COA_KEY \
--rate-limit=9999999 \
--rpc-host=0.0.0.0 \

> --gas-price=$GAS_PRICE

Author: ```suggestion
--gas-price=$GAS_PRICE \
--tx-state-validation=local-index
```

using the same release version. We are currently working on an export/import feature that will enable gateway operators to 
store state snapshots to bootstrap newly created nodes without the delay.

:::note

If you are upgrading the gateway from pre-v1.0.0 release versions the indexed data directory will need to be reindexed from genesis. 
You will not be able to re-use the DB data dir from the previous versions. 

:::

### Account and Key Management

EVM Gateway allows for Google and AWS Key Management Service (KMS) setup, which is the recommended way of setting up the gateway 
for live networks. We recommend creating multiple KMS keys for the same Flow account (ideally 10 or more), how many depends on 
the desired transaction throughput, since the keys are used in rotation when submitting the transactions. If too few keys 
are configured it may result in sequence number collisions if the same key is used concurrently by multiple EVM client requests. 

### KMS Configuration
```
--coa-cloud-kms-project-id=your-project-kms-id \
--coa-cloud-kms-location-id=global \
--coa-cloud-kms-key-ring-id=your-project-kms-key-ring-id \

> --coa-cloud-kms-keys=example-gcp-kms1@1,example-gcp-kms2@1 \

Author: ```suggestion
--coa-cloud-kms-key=example-gcp-kms1@1,example-gcp-kms2@1 \
```
